File: Editorial/39_Combination_Sum/Go/1_Backtracking.go
func combinationSum(candidates []int, target int) [][]int {
    var results [][]int
    var comb []int
    backtrack(target, comb, 0, candidates, &results)
    return results
}

func backtrack(
    remain int,
    comb []int,
    start int,
    candidates []int,
    results *[][]int,
) {
    if remain == 0 {
        // Make a deep copy of the current combination to add to results
        newComb := make([]int, len(comb))
        copy(newComb, comb)
        *results = append(*results, newComb)
        return
    } else if remain < 0 {
        // Exceed the scope, stop exploration.
        return
    }

    for i := start; i < len(candidates); i++ {
        // Add the number into the combination
        comb = append(comb, candidates[i])
        // Continue to explore with the updated combination
        backtrack(remain-candidates[i], comb, i, candidates, results)
        // Backtrack, remove the number from the combination
        comb = comb[:len(comb)-1]
    }
}

File: Editorial/39_Combination_Sum/CPP/1_Backtracking.cpp
class Solution {
public:
    void backtrack(int remain, vector<int>& comb, int start,
                   const vector<int>& candidates,
                   vector<vector<int>>& results) {
        if (remain == 0) {
            // make a deep copy of the current combination
            results.push_back(comb);
            return;
        } else if (remain < 0) {
            // exceed the scope, stop exploration.
            return;
        }

        for (int i = start; i < candidates.size(); ++i) {
            // add the number into the combination
            comb.push_back(candidates[i]);
            backtrack(remain - candidates[i], comb,
                      i,  // not i + 1 because we can reuse same elements
                      candidates, results);
            // backtrack, remove the number from the combination
            comb.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> results;
        vector<int> comb;

        backtrack(target, comb, 0, candidates, results);
        return results;
    }
};
